### ISSUE TITLE
Generating implementation in the same folder

### ISSUE THREAD

**Issue by lmichaudel** [eriks47/generate.nvim#1]
This plugin works when header and source are on the same folder.
When header is in include/ and source in src/ the command will generate a new source file inside include/ instead of writing in the one located in src/

**Comment by eriks47** [comment]
Yeah, currently the logic is hard-coded, the appropriate solution would probably be to use the user provided configuration. However, then the configuration would have to be changed for every project which may be annoying. Currently it looks like the best option is to look for a source file directory if the current directory is named "Include" or anything like that, because as far as I know that's a common pattern. I'll write a patch.

**Comment by eriks47** [comment]
Now if the header file is in a directory called "inc" or "include" (case insensitive), it looks if the directory that contains the include directory also happens to have a directory called "src" or "source", if that's the case the source file is placed there. Does the update resolve the issue?

**Comment by lmichaudel** [comment]
Now writing to the correct file, but I have a new issue, let me know if you want me to open a new issue, but when the source file already contains some code; instead of appending the generated function, it erase all code and only append what was not present.
Example : if I have two function foo() and bar() but only foo() is implemented, then it will erase the foo() implementation and only generate a bar() implementation.
Let me know if you are able to replicate this

**Comment by eriks47** [comment]
I'm not sure I can replicate this. Does the plugin really delete anything? Because it seems like it only appends content. Running `rg fs_open` on the source tree shows me that all writes are done in append mode
```
$ rg fs_open
generate/source.lua
110:  local fd = uv.fs_open(path, 'a', 438)

generate/filesystem.lua
101:  local fd = uv.fs_open(path, 'a', 438)
```

**Comment by lmichaudel** [comment]
I'm subject to a weird behaviour, certainly caused by something else, i'll link a screen capture 

https://github.com/eriks47/generate.nvim/assets/38987429/bb3c24d5-548e-4127-978f-5bc4aed4307b

**Comment by lmichaudel** [comment]
Also regarding the first issue mentionned we could use the same system as [Ouroboros](https://github.com/jakemason/ouroboros.nvim) to effeciently find corresponding source/header files

**Comment by eriks47** [comment]
So if I understand correctly (correct me if I'm wrong), the engine.cpp file you opened initially is the same one you opened later? Or is it a different file? And was the file blank and its contents generated, or was already filled with implementations?

**Comment by lmichaudel** [comment]
Yes it's the same file; which was empty at start. Then I ran the command, it showed me the cpp file automatically with only the header; then I closed my buffer and re-opened it

**Comment by eriks47** [comment]
Could you check if running `:edit` or `:redraw` fixes the "glitched" buffer?

**Comment by lmichaudel** [comment]
It doesn't, I will try on a cleaner instance of neovim

**Comment by eriks47** [comment]
Does the latest commit solve the issue?

**Comment by lmichaudel** [comment]
Nope, still not working :/ 
It only write the first line (the include of the header); then 4 empty lines

**Comment by lmichaudel** [comment]
HI, i'm currently on vacation, but I did test with a clean neovim install and the same behaviour happens.
Once back home i'll open a new issue and I will close this one as it's not related.

More informations : I'm on a macbook air m2 with latest version of Ventura, i'm using iTerm2 and latest stable neovim

**Comment by Sologala** [comment]
> Yeah, currently the logic is hard-coded, the appropriate solution would probably be to use the user provided configuration. However, then the configuration would have to be changed for every project which may be annoying. Currently it looks like the best option is to look for a source file directory if the current directory is named "Include" or anything like that, because as far as I know that's a common pattern. I'll write a patch.

@eriks47 Hi, I think hard-coded rules cannot be robust for different codebase organizing styles. Lsp provides the ability to get the implementation file path. I submitted a [PR](https://github.com/eriks47/generate.nvim/pull/6) which realize it.
